File: greedy/3219-minimum-cost-for-cutting-cake-ii.py
from typing import List
import unittest

# https://leetcode.com/problems/minimum-cost-for-cutting-cake-ii/

# python3 -m unittest greedy/3219-minimum-cost-for-cutting-cake-ii.py


class Solution(unittest.TestCase):
    def minimumCost(self, m: int, n: int, horizontalCut: List[int], verticalCut: List[int]) -> int:
        horizontalCut.sort(reverse=True)
        verticalCut.sort(reverse=True)
        
        m, n = m-1, n-1
        total, hid, vid = 0, 0, 0
        
        while hid < m and vid < n:
            if horizontalCut[hid] >= verticalCut[vid]:
                total += horizontalCut[hid] * (vid+1)
                hid += 1
            else:
                total += verticalCut[vid] * (hid+1)
                vid += 1

        if hid < m:
            total += sum(horizontalCut[hid:]) * (vid+1)
        elif vid < n:
            total += sum(verticalCut[vid:]) * (hid+1)

        return total

    def test(self):
        for m, n, horizontalCut, verticalCut, expected in [
            (3, 2, [1,3], [5], 13),
            (2, 2, [7], [4], 15),
        ]:
            output = self.minimumCost(m, n, horizontalCut, verticalCut)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: greedy/3219-minimum-cost-for-cutting-cake-ii.go
package greedy

import "sort"

// https://leetcode.com/problems/minimum-cost-for-cutting-cake-ii/

func minimumCost(m int, n int, horizontalCut []int, verticalCut []int) int64 {
	sort.Ints(horizontalCut)
	sort.Ints(verticalCut)

	hid, vid := m-2, n-2
	hpieces, vpieces := 1, 1

	var total int64
	for hid >= 0 && vid >= 0 {
		if horizontalCut[hid] > verticalCut[vid] {
			total += int64(horizontalCut[hid] * vpieces)
			hid--
			hpieces++
		} else {
			total += int64(verticalCut[vid] * hpieces)
			vid--
			vpieces++
		}
	}

	for ; hid >= 0; hid-- {
		total += int64(horizontalCut[hid] * vpieces)
	}
	for ; vid >= 0; vid-- {
		total += int64(verticalCut[vid] * hpieces)
	}

	return total
}
